stdlib/public/runtime/SectionDataStatic.cpp

```diff
@@ -0,0 +1,61 @@
+//===-- SectionDataStatic.cpp -----------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of functions to read data sections from static executable.
+//
+//===----------------------------------------------------------------------===//
+
+#include <cstdlib>
+#include <cstring>
+#include "swift/Runtime/Debug.h"
+#include "SectionData.h"
+
+
+#if defined SUPPORTS_STATIC_BINARIES
+
+// If creating a static binary using --static-executable,
+// gold will set these to the data sections via --defsym
+const void *__swift2_protocol_conformances_start = nullptr;
+const void *__swift2_type_metadata_start = nullptr;
+
+
+void
+swift::_swift_initializeCallbacksForSectionData(InspectArgs *inspectArgs) {
+   const char *name = inspectArgs->sectionName;
+   const void **sectionDataAddr = nullptr;
+
+   if (!strcmp(name, SWIFT_PROTOCOL_CONFORMANCES_SECTION)) {
+     sectionDataAddr = &__swift2_protocol_conformances_start;
+   } else if (!strcmp(name, SWIFT_TYPE_METADATA_SECTION)) {
+     sectionDataAddr = &__swift2_type_metadata_start;
+   } else {
+     fatalError(/* flags = */ 0, "Invalid section name: %s", name);
+   }
+
+   if (sectionDataAddr == nullptr || *sectionDataAddr == nullptr) {
+     fatalError(/* flags = */ 0, "Static data section `%s' is null", name);
+   }
+   auto blockAddr = reinterpret_cast<const uint8_t*>(sectionDataAddr);
+   auto blockSize = *reinterpret_cast<const uint64_t*>(blockAddr);
+   blockAddr += sizeof(blockSize);
+   inspectArgs->fnAddImageBlock(blockAddr, blockSize);
+}
+
+extern "C" int
+dladdr(const void* addr, Dl_info* info)
+{
+  return 0;
+}
```

Comment:
Thanks @spevans, this is shaping up nicely! I took a shot at refactoring all the image inspection code in #5846. It should be a lot easier to cleanly slot in an `ImageInspectionStatic.cpp` file once that goes in; all you'd need is something like:

```
extern const char Conformances __asm__(".swift2_protocol_conformances_start");
extern const char TypeMetadataRecords __asm__(".swift2_type_metadata_start");

void swift::initializeProtocolConformanceLookup() {
  // The image is statically linked, so we can just directly reference the symbol
  // that signals the beginning of the section.
  // ld should mash all the conformance records together for us, and we don't need
  // to be concerned about dynamic loading.
  uint64_t size;
  memcpy(&size, Conformances, sizeof(size));
  addImageProtocolConformanceBlockCallBack(Conformances + sizeof(size), size);
}
```

and something similar for type metadata records.